README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

A simple pin text box that supports customization of cursor, border, font, etc

## Getting started

read install tab

## Usage

object

```dart
PinTextField PinTextField({
  Key? key,
  int count = 6,
  double height = 42.0,
  double width = 42.0,
  Color cursorColor = const Color.fromRGBO(255, 96, 10, 1),
  Color fontColor = const Color.fromRGBO(0, 0, 0, 1),
  double fontSize = 16,
  double borderWidth = 1.0,
  Color borderColor = const Color.fromRGBO(204, 204, 204, 1),
  double focusBorderWidth = 1.0,
  Color focusBorderColor = const Color.fromRGBO(100, 105, 110, 1),PinTextFieldBorderType borderType = PinTextFieldBorderType.Underline,
  String finishedStringSeparator = "",
  void Function(String)? onFinished,
})
```

usage

```dart
PinTextField(
    borderType: PinTextFieldBorderType.Underline,
    finishedStringSeparator: ",",
    onFinished: (value) {
        print(value);
    })
```

![example](./example/example.png)


CHANGELOG.md
## 1.0.0

* init package


test/pin_text_field_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:pin_text_field/pin_text_field.dart';


test/flutter_pin_text_field.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_pin_text_field/flutter_pin_text_field.dart';


lib/pin_text_field.dart
library pin_text_field;

import 'package:flutter/material.dart';

enum PinTextFieldBorderType { Underline, Outline }

class PinTextFieldItem extends StatefulWidget {
  double height;
  double width;
  Color cursorColor;
  Color fontColor;
  double fontSize;
  Color borderColor;
  Color focusBorderColor;
  double borderWidth;
  double focusBorderWidth;
  PinTextFieldBorderType borderType;
  void Function(String, FocusNode)? onChanged;
  int index;
  PinTextFieldItem(
      {Key? key,
      required this.index,
      required this.cursorColor,
      required this.height,
      required this.width,
      required this.fontColor,
      required this.fontSize,
      required this.borderColor,
      required this.focusBorderColor,
      required this.borderWidth,
      required this.focusBorderWidth,
      required this.borderType,
      this.onChanged})
      : super(key: key);
  @override
  State<PinTextFieldItem> createState() => _PinTextFieldItemState();
}

class _PinTextFieldItemState extends State<PinTextFieldItem> {
  late TextEditingController controller;
  FocusNode focusNode = FocusNode();
  @override
  void initState() {
    super.initState();
    controller = TextEditingController(text: "");
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: widget.height,
      width: widget.width,
      child: TextField(
        focusNode: focusNode,
        onChanged: ((value) {
          if (widget.onChanged != null) {
            widget.onChanged!(value, focusNode);
          }
        }),
        maxLength: 1,
        keyboardType: TextInputType.number,
        cursorColor: widget.cursorColor,
        decoration: InputDecoration(
          contentPadding: const EdgeInsets.all(0),
          counterText: "",
          enabledBorder: widget.borderType == PinTextFieldBorderType.Underline
              ? UnderlineInputBorder(
                  borderSide: BorderSide(
                      color: widget.borderColor, width: widget.borderWidth),
                )
              : OutlineInputBorder(
                  borderSide: BorderSide(
                      color: widget.borderColor, width: widget.borderWidth)),
          focusedBorder: widget.borderType == PinTextFieldBorderType.Underline
              ? UnderlineInputBorder(
                  borderSide: BorderSide(
                      color: widget.focusBorderColor,
                      width: widget.focusBorderWidth),
                )
              : OutlineInputBorder(
                  borderSide: BorderSide(
                      color: widget.focusBorderColor,
                      width: widget.focusBorderWidth)),
        ),
        textAlign: TextAlign.center,
        controller: controller,
        style: TextStyle(color: widget.fontColor, fontSize: widget.fontSize),
      ),
    );
  }
}

class PinTextField extends StatefulWidget {
  int count;
  double height;
  double width;
  Color cursorColor;
  Color fontColor;
  double fontSize;
  double borderWidth;
  Color borderColor;
  double focusBorderWidth;
  Color focusBorderColor;
  PinTextFieldBorderType borderType;
  String finishedStringSeparator;
  void Function(String)? onFinished;
  PinTextField(
      {Key? key,
      this.count = 6,
      this.height = 42.0,
      this.width = 42.0,
      this.cursorColor = const Color.fromRGBO(255, 96, 10, 1),
      this.fontColor = const Color.fromRGBO(0, 0, 0, 1),
      this.fontSize = 16,
      this.borderWidth = 1.0,
      this.borderColor = const Color.fromRGBO(204, 204, 204, 1),
      this.focusBorderWidth = 1.0,
      this.focusBorderColor = const Color.fromRGBO(100, 105, 110, 1),
      this.borderType = PinTextFieldBorderType.Underline,
      this.finishedStringSeparator = "",
      this.onFinished})
      : super(key: key);
  @override
  State<PinTextField> createState() => _PinTextFieldState();
}

class _PinTextFieldState extends State<PinTextField> {
  List<int?> intList = [];
  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    intList = List.generate(widget.count, (index) => null);
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: List.generate(
              widget.count,
              (index) => PinTextFieldItem(
                    index: index,
                    borderWidth: widget.borderWidth,
                    focusBorderWidth: widget.focusBorderWidth,
                    cursorColor: widget.cursorColor,
                    height: widget.height,
                    width: widget.width,
                    focusBorderColor: widget.focusBorderColor,
                    borderColor: widget.borderColor,
                    fontColor: widget.fontColor,
                    borderType: widget.borderType,
                    fontSize: widget.fontSize,
                    onChanged: (value, focusNode) {
                      if (value != "") {
                        intList[index] = int.parse(value);
                        if (index < widget.count - 1) {
                          focusNode.nextFocus();
                        }
                      } else {
                        intList[index] = null;
                        if (index > 0) {
                          focusNode.previousFocus();
                        }
                      }
                      var array =
                          intList.where((element) => element != null).toList();
                      if (array.length == widget.count) {
                        if (widget.onFinished != null) {
                          var numberList = array.map((e) => e ?? 0).toList();
                          var string =
                              numberList.join(widget.finishedStringSeparator);
                          widget.onFinished!(string);
                          focusNode.unfocus();
                        }
                      }
                    },
                  ))),
    );
  }
}


lib/flutter_pin_text_field.dart
library flutter_pin_text_field;

import 'package:flutter/material.dart';

enum PinTextFieldBorderType { Underline, Outline }

class PinTextFieldItem extends StatefulWidget {
  double height;
  double width;
  Color cursorColor;
  Color fontColor;
  double fontSize;
  Color borderColor;
  Color focusBorderColor;
  double borderWidth;
  double focusBorderWidth;
  PinTextFieldBorderType borderType;
  void Function(String, FocusNode)? onChanged;
  int index;
  PinTextFieldItem(
      {Key? key,
      required this.index,
      required this.cursorColor,
      required this.height,
      required this.width,
      required this.fontColor,
      required this.fontSize,
      required this.borderColor,
      required this.focusBorderColor,
      required this.borderWidth,
      required this.focusBorderWidth,
      required this.borderType,
      this.onChanged})
      : super(key: key);
  @override
  State<PinTextFieldItem> createState() => _PinTextFieldItemState();
}

class _PinTextFieldItemState extends State<PinTextFieldItem> {
  late TextEditingController controller;
  FocusNode focusNode = FocusNode();
  @override
  void initState() {
    super.initState();
    controller = TextEditingController(text: "");
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: widget.height,
      width: widget.width,
      child: TextField(
        focusNode: focusNode,
        onChanged: ((value) {
          if (widget.onChanged != null) {
            widget.onChanged!(value, focusNode);
          }
        }),
        maxLength: 1,
        keyboardType: TextInputType.number,
        cursorColor: widget.cursorColor,
        decoration: InputDecoration(
          contentPadding: const EdgeInsets.all(0),
          counterText: "",
          enabledBorder: widget.borderType == PinTextFieldBorderType.Underline
              ? UnderlineInputBorder(
                  borderSide: BorderSide(
                      color: widget.borderColor, width: widget.borderWidth),
                )
              : OutlineInputBorder(
                  borderSide: BorderSide(
                      color: widget.borderColor, width: widget.borderWidth)),
          focusedBorder: widget.borderType == PinTextFieldBorderType.Underline
              ? UnderlineInputBorder(
                  borderSide: BorderSide(
                      color: widget.focusBorderColor,
                      width: widget.focusBorderWidth),
                )
              : OutlineInputBorder(
                  borderSide: BorderSide(
                      color: widget.focusBorderColor,
                      width: widget.focusBorderWidth)),
        ),
        textAlign: TextAlign.center,
        controller: controller,
        style: TextStyle(color: widget.fontColor, fontSize: widget.fontSize),
      ),
    );
  }
}

class PinTextField extends StatefulWidget {
  int count;
  double height;
  double width;
  Color cursorColor;
  Color fontColor;
  double fontSize;
  double borderWidth;
  Color borderColor;
  double focusBorderWidth;
  Color focusBorderColor;
  PinTextFieldBorderType borderType;
  String finishedStringSeparator;
  void Function(String)? onFinished;
  PinTextField(
      {Key? key,
      this.count = 6,
      this.height = 42.0,
      this.width = 42.0,
      this.cursorColor = const Color.fromRGBO(255, 96, 10, 1),
      this.fontColor = const Color.fromRGBO(0, 0, 0, 1),
      this.fontSize = 16,
      this.borderWidth = 1.0,
      this.borderColor = const Color.fromRGBO(204, 204, 204, 1),
      this.focusBorderWidth = 1.0,
      this.focusBorderColor = const Color.fromRGBO(100, 105, 110, 1),
      this.borderType = PinTextFieldBorderType.Underline,
      this.finishedStringSeparator = "",
      this.onFinished})
      : super(key: key);
  @override
  State<PinTextField> createState() => _PinTextFieldState();
}

class _PinTextFieldState extends State<PinTextField> {
  List<int?> intList = [];
  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    intList = List.generate(widget.count, (index) => null);
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: List.generate(
              widget.count,
              (index) => PinTextFieldItem(
                    index: index,
                    borderWidth: widget.borderWidth,
                    focusBorderWidth: widget.focusBorderWidth,
                    cursorColor: widget.cursorColor,
                    height: widget.height,
                    width: widget.width,
                    focusBorderColor: widget.focusBorderColor,
                    borderColor: widget.borderColor,
                    fontColor: widget.fontColor,
                    borderType: widget.borderType,
                    fontSize: widget.fontSize,
                    onChanged: (value, focusNode) {
                      if (value != "") {
                        intList[index] = int.parse(value);
                        if (index < widget.count - 1) {
                          focusNode.nextFocus();
                        }
                      } else {
                        intList[index] = null;
                        if (index > 0) {
                          focusNode.previousFocus();
                        }
                      }
                      var array =
                          intList.where((element) => element != null).toList();
                      if (array.length == widget.count) {
                        if (widget.onFinished != null) {
                          var numberList = array.map((e) => e ?? 0).toList();
                          var string =
                              numberList.join(widget.finishedStringSeparator);
                          widget.onFinished!(string);
                          focusNode.unfocus();
                        }
                      }
                    },
                  ))),
    );
  }
}
